packages/dd-trace/src/aiguard/sdk.js

```diff
@@ -77,20 +78,33 @@ class AIGuard extends NoopAIGuard {
     this.#initialized = true
   }
 
-  #truncate (messages) {
+  /**
+   * Returns a safe copy of the messages to be serialized into the meta struct.
+   *
+   * - Clones each message so callers cannot mutate the data set in the meta struct.
+   * - Truncates the list of messages and `content` fields emitting metrics accordingly.
+   */
+  #buildMessagesForMetaStruct (messages) {
     const size = Math.min(messages.length, this.#maxMessagesLength)
     if (messages.length > size) {
       appsecMetrics.count(AI_GUARD_TELEMETRY_TRUNCATED, { type: 'messages' }).inc(1)
     }
-    const result = messages.slice(-size)
-
+    const result = []
     let contentTruncated = false
-    for (let i = 0; i < size; i++) {
-      const message = result[i]
+    for (let i = messages.length - size; i < messages.length; i++) {
+      const message = { ...messages[i] }
       if (message.content?.length > this.#maxContentSize) {
         contentTruncated = true
-        result[i] = { ...message, content: message.content.slice(0, this.#maxContentSize) }
+        message.content = message.content.slice(0, this.#maxContentSize)
+      }
+      if ('tool_calls' in message) {
+        // deep copy
+        message.tool_calls = message.tool_calls.map(call => ({
+          id: call.id,
+          function: call.function ? { ...call.function } : null,
+        }))
       }
```

Comment:
```suggestion
      message.tool_calls = rfdc(message.tool_calls)
```

The library is already included and could be used for deep copying :)